rfd/0147-repo-fork-syncing.md

```diff
@@ -0,0 +1,60 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+ 
+# RFD 144 - Automatic Repo Syncing
+ 
+## What
+ 
+This RFD describes how we can sync a repo continuously and in an automatic way.  This can be used to provide private copies of a repo.
+ 
+## Why
+ 
+We currently face a few frictions which could be addressed through a more robust repo sync mechanism:
+1. Our internal `teleport-private` fork used for security development must be manually updated
+2. Some tooling (Dependabot) does not allow scanning for non-default branches, necessitating manual efforts to manage our release branches
+3. Because much of the CI and scanning configuration is committed into the repo, we face noise from duplicate alerts when we use a repo for testing or as an internal fork
+ 
+## Details
+ 
+As such we need a solution that can provide the following:
+* Automatic and regular updates to a repo to ensure it's in sync
+* The ability to apply changes to the upstream repo (for example disable Actions that we don't want running on the repo copy, or enable additional scanning / actions)
+ 
+### Branch Structure
+ 
+* `main` - This will be the repo's default branch.  This branch will contain the upstream history in addition to commits needed for the sync process or other custom changes on the repo copy
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch.  If there is a need to create a PR in this repo the history should be based off this rather than `main`
+* `sync/rebase` - This branch is where our custom changes will be committed.  It will be rebased on to the upstream changes before being committed into `main`
+ 
+#### Branch Protections
+ 
+The only branch that will be protected is the `sync/rebase` branch.  Other branches are expected to allow force pushes so they can be forcefully synced based off the upstream state.
+ 
+### High Level Implementation
+ 
+At a high level there are two primary components to this implementation:
+1. A GitHub action which will be scheduled as well as triggered on changes to the `sync/rebase` branch.  This action will perform the actual rebase action and commit the results to the respective branches above.
+2. A script that is invoked when the rebase can not be applied cleanly.  This can allow some custom rebase logic as a means to reduce how often the `sync/rebase` branch needs to be manually updated to avoid rebase conflicts.  A simple example is that workflow automations not desired in the repo copy may be removed in the `sync/rebase` branch.  That means that any changes to those committed workflows would result in a conflict.  However it can be safely resolved that any modified files removed in the `sync/rebase` branch can be resolved by simply removing.
```

Comment:
> However it can be safely resolved that any modified files removed in the `sync/rebase` branch can be resolved by simply removing

Suggestion: We could have a list of "ignored" files, so we clearly spell out somewhere what we removed why it is safe for the script to forgo those. Also we can just commit that as a file to the branch itself.

Reply:
Why have this outside of the commit history itself?  Right now the explicit removal of a file in the `rebase` branch is effectively marking that file as "ignored"

Reply:
Because in some cases we don't really care about (eg, action files), but in other cases removed files could have meaningful upstream changes.

The suggestion would keep a versioned list in the repo, as well as the actual delete of the file, so everything is explicit and tracked by the VCS.

Reply:
In theory any custom logic could be provided in the resolution script.  I suggest we keep the RFD generic, and if we have a need for building an ignore list, we update the resolution script to accommodate that repos needs.  Thoughts?

Reply:
Sure, we don't need to go overboard now. Still, I'd rather we limited, by design, the auto-resolution of deletions to only certain folders to begin with, like .github.